tableauserverclient/helpers/strings.py

```diff
@@ -0,0 +1,66 @@
+from functools import singledispatch
+from typing import TypeVar
+import requests
+import sys
+
+
+# the redact method can handle either strings or bytes, but it can't mix them.
+# Generic type so we can write the actual logic once, then use singledispatch to
+# create the replacement text with the correct type
+T = TypeVar("T", str, bytes)
+
+
+# TODO: ideally this would be in the logging config
+def safe_to_log(server_response: requests.Response) -> str:
+    """Checks if the server_response content is not xml (eg binary image or zip)
+    and replaces it with a constant"""
+    ALLOWED_CONTENT_TYPES = ("application/xml", "application/xml;charset=utf-8")
+    if server_response.headers.get("Content-Type", None) not in ALLOWED_CONTENT_TYPES:
+        return "[Truncated File Contents]"
+
+    """ Check to determine if the response is a text response (xml or otherwise)
+        so that we do not attempt to log bytes and other binary data. """
+    if not server_response.content or not server_response.encoding:
+        return ""
+    # max length 1000
+    loggable_response: str = server_response.content.decode(server_response.encoding)[:1000]
+    redacted_response: str = redact(loggable_response)
+    return redacted_response
+
+
+def _replace(text: T, position: int, replacement: T) -> T:
+    result: T = text[:position] + replacement + text[position + len(replacement) :]
+    return result
+
+
+def _redact_typeful(content: T, target: T, replacement: T) -> T:
+    search_start: int = 0
+    while search_start >= 0:
+        try:
+            replacement_begin: int = content.index(target, search_start) + 8
+            content = _replace(content, replacement_begin, replacement)
+            search_start = replacement_begin + 8
+        except ValueError:
+            search_start = -1
+    content = content.replace(target, replacement)
+    return content
```

Comment:
how is this supposed to work?
I fail to see how this reliably removes passwords from the string

Reply:
The current problem is that we are logging the xml of a datasource or workbook, and including the password for data connections if it is embedded. This shows up in the format password=blahblah, so we are redacting both the word 'password' and the text immediately *after* the word. 
I haven't written it to be more specific or do any xml handling because I want it to be able to apply to whatever string is being logged, without us having to identify all the attributes that we use for passwords.

Reply:
Thanks for the high-level explanation! I agree with the general approach.
The implementation, however, does not do what you are describing...

It only partially replaces the text after the word.

If I call
```
_redact_typeful("password=my_super_secret_passphrase_which_nobody_should_ever_see", "password", "redacted")
```
I get
> redactedredactedr_secret_passphrase_which_nobody_should_ever_see

Note how a significant part of my password is still visible

Reply:
True. I've added the ability to keep going until the first quote or xml end-bracket character. This will still leave part of some passwords visible, but seems like a decent trade-off?

Reply:
removed this logic altogether